### PR TITLE
Implement lazy Property for Cfgu Declaration

### LINKED ISSUE #335: Implement `lazy` Property for Cfgu Declaration
### Suggestion

Adding a `lazy` boolean property to the Cfgu declaration. This property allows users to specify that a value for a given `ConfigKey` should only be set during the `eval` command execution, either using the `configs` parameter or the `--config` flag in the CLI. Essentially, `lazy: true` implies the following:
- Values for this key cannot be assigned using the `upsert` command.
- Any existing value (`Config`) for this key in any store will be disregarded during `eval`.
- The only method to assign a value to this key is through the `eval` command with the `configs` parameter.
- If marked as `required`, the user must provide a value for this key during `eval`.
- This property cannot be combined with `template` and `default` properties.

```json
# my_schema.cfgu.json
{
  ...
  "MY_EMAIL": {
    "type": "Email",
    "lazy": true
  }
  ...
} 
```

```bash
configu upsert ... --schema './my_schema.cfgu.json' -c "MY_EMAIL=ran@configu.com" # Error

configu eval --schema './my_schema.cfgu.json' -c "MY_EMAIL=ran@configu.com" # OK
```

### Motivation

This feature enhances the Cfgu declaration's versatility, catering to scenarios where configuration values need to be dynamic or context-specific. It offers greater control over when and how configuration values are assigned, particularly beneficial in environments with varying runtime contexts. By enabling lazy evaluation of config values, this feature facilitates more adaptable and responsive configuration setups, particularly useful in complex, multi-stage deployment processes or when dealing with sensitive information that requires contextual assignment.

### Context

_No response_

### PR DESCRIPTION
**Pull request by RonConfigu**
## QA Report I
- [x] Test Units Passes

Closes #335

### CODE REVIEW

File: types/Cfgu.ts [hunk 1]
```diff
@@ -1,40 +1,40 @@
 export type CfguType =
```
Comment (RichardAkman) [inline types/Cfgu.ts:1]: I'm guessing your IDE did this? Let's not change this 🙏🏻

File: ts/packages/ts/src/commands/UpsertCommand.ts [hunk 2]
```diff
@@ -60,6 +60,10 @@ export class UpsertCommand extends Command<void> {
           );
         }
 
+        if (cfgu.lazy) {
+          throw new ConfigError('invalid config key', `key "${key}" is lazy and cannot be set directly`, errorScope);
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/UpsertCommand.ts:64]: How about this to keep it similar to the template error:
```suggestion
          throw new ConfigError('invalid config key', `keys declared as lazy mustn't have a value`, errorScope);
```

File: ts/packages/ts/src/commands/EvalCommand.ts [hunk 3]
```diff
@@ -36,7 +36,14 @@ export class EvalCommand extends Command<EvalCommandReturn> {
   }
 
   private evalFromConfigsOverride(result: EvalCommandReturn): EvalCommandReturn {
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/EvalCommand.ts:38]: I strongly feel like this method should remain untouched and `validateResult` should be responsible for throwing the `required` error which it already does. Is there any specific reason why we need to duplicate the validation here?

File: ts/packages/ts/src/commands/commands.test.ts [hunk 4]
```diff
@@ -221,5 +243,147 @@ describe(`commands`, () => {
         await Promise.all(deletePromises);
       }
     });
+    describe('Tests for lazy configs', () => {
+      test('run EvalCommand without configs but  `cfgu.lazy && cfgu.required = true`', async () => {
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:247]:
```suggestion
      test('run EvalCommand without configs but `cfgu.lazy && cfgu.required = true`', async () => {
```

File: ts/packages/ts/src/commands/commands.test.ts [hunk 5]
```diff
@@ -221,5 +243,147 @@ describe(`commands`, () => {
...
+        expect.assertions(1);
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:248]: Not sure why we need these, please explain 🙏🏻

File: ts/packages/ts/src/commands/commands.test.ts [hunk 6]
```diff
@@ -221,5 +243,147 @@ describe(`commands`, () => {
...
+        await expect(() =>
+          new EvalCommand({
+            store: store1,
+            set: set1,
+            schema: new ConfigSchema('lazy', {
+              K1: {
+                type: 'String',
+                lazy: true,
+                required: true,
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:295]: the test description says `cfgu.required = false`

File: ts/packages/ts/src/commands/commands.test.ts [hunk 7]
```diff
@@ -221,5 +243,147 @@ describe(`commands`, () => {
...
+              required: true,
+            },
+          }),
+          configs: {
+            K1: '1',
+          },
+        }).run();
+        expect(result).toMatchObject({ K1: { result: { value: '1' } } });
+      });
+      test('run EvalCommand without override for lazy config when `cfgu.required = false`', async () => {
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:341]: the `required` prop is not defined, we need to clarify the test names and have them align with what's going on inside the test.

File: ts/packages/ts/src/commands/commands.test.ts [hunk 8]
```diff
@@ -221,5 +243,147 @@ describe(`commands`, () => {
...
+              type: 'String',
+              lazy: true,
+            },
+          }),
+        }).run();
+        expect(result).toMatchObject({ K1: { result: { value: '' } } });
+      });
+
+      test('run EvalCommand with override for lazy config when there is a value in the store and get the override value in the result', async () => {
+        await store1.set([{ set: set1.path, key: 'K1', value: '2' }]);
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/commands.test.ts:356]: I think we need to avoid any direct use of the store methods from outside the commands themselves, we don't want people to get the idea that we want them to do these things, even for tests. Prefer using upsert.

**Review by RichardAkman — CHANGES_REQUESTED** [review]
Great work 💯 please review my comments 🙏🏻

File: types/Cfgu.ts (hunk 1 above)
Comment (RonConfigu) [reply types/Cfgu.ts:1]: Sure. sorry

File: ts/packages/ts/src/commands/commands.test.ts
[the remark this answers is not in the document]
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:1]: There are no other PR's. Please elaborate on the problem here.
Thanks

File: ts/packages/ts/src/commands/commands.test.ts (hunk 8 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:356]: Strongly disagree.
This test has nothing to do we UpsertCommand. If it would be possible I would imitate a store with these in it.
Generally speaking, in uint-tests like in real life, one should eliminate all variables from the experiment and only change the tested one. proving two variables in one test is not valid for either. (unless they are entangled)

File: ts/packages/ts/src/commands/commands.test.ts (hunk 6 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:295]: The "error" here is in the title not the prop.
the test right below has this prop = false.

Thanks

File: ts/packages/ts/src/commands/commands.test.ts (hunk 7 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts]: all tests names have been refactored. please review.

File: ts/packages/ts/src/commands/commands.test.ts
[the remark this answers is not in the document]
Comment (RichardAkman) [reply ts/packages/ts/src/commands/commands.test.ts:1]: This code is duplicated 6 times:
```node.js
const evalResult = await new EvalCommand({
        store: store1,
        set: set1,
        schema: new ConfigSchema('mutate', {
          KEY0: {
            type: 'String',
          },
          KEY1: {
            type: 'String',
          },
        }),
        configs: {
          KEY0: 'KEY0',
          KEY1: 'KEY1',
        },
      }).run();
```

let's keep things DRY and recycle this code.

File: ts/packages/ts/src/commands/commands.test.ts (hunk 8 above)
Comment (RichardAkman) [reply ts/packages/ts/src/commands/commands.test.ts:333]: Then at the very least remove the setup part from the test itself by using something like [`beforeAll`](https://jestjs.io/docs/setup-teardown#one-time-setup)

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 3 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/EvalCommand.ts]: done

File: ts/packages/ts/src/commands/EvalCommand.ts [hunk 9]
```diff
@@ -291,7 +291,10 @@ export class EvalCommand extends Command<EvalCommandReturn> {
       ...(await this.evalFromStoreSet(
         _.pickBy(
           result,
-          (current) => current.result.origin === EvaluatedConfigOrigin.EmptyValue && !current.context.cfgu.template,
+          (current) =>
+            current.result.origin === EvaluatedConfigOrigin.EmptyValue &&
+            !current.context.cfgu.template &&
+            !current.context.cfgu.lazy,
```
Comment (rannn505) [inline ts/packages/ts/src/commands/EvalCommand.ts:297]: this should already been done on the `evalFromConfigsOverride` no?
you should validate somewhere that all lazy configs are set via the config override.

**Review by rannn505 — CHANGES_REQUESTED** [review]
@RonConfigu great work

File: ts/packages/ts/src/commands/EvalCommand.ts [hunk 10]
```diff
@@ -158,6 +171,7 @@ export class EvalCommand extends Command<EvalCommandReturn> {
 
     const mergedResults = _([pipe, result])
       .flatMap((current) => _.values(current))
+      .value()
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/EvalCommand.ts:174]: I remember that you had a linting issue here but it's an unnecessary change that adds more complexity to the operation and is irrelevant to this PR. Let's omit this for now 🙏🏻

File: ts/packages/ts/src/commands/EvalCommand.ts [hunk 11]
```diff
@@ -37,7 +37,20 @@ export class EvalCommand extends Command<EvalCommandReturn> {
 
   private evalFromConfigsOverride(result: EvalCommandReturn): EvalCommandReturn {
     if (!this.parameters.configs) {
-      return result;
+      return _.mapValues(result, (current) => {
```
Comment (RichardAkman) [inline ts/packages/ts/src/commands/EvalCommand.ts:40]: The condition here is incorrect, if you have two lazy keys and one is provided and the second isn't you'll end up evaluating from the store for the missing key.

How about this direction? We let the existing override logic always run and check if there is a lazy property on the keys
```node.js
  private evalFromConfigsOverride(result: EvalCommandReturn): EvalCommandReturn {
    return _.mapValues(result, (current) => {
      const { context } = current;
      const hasConfigOverrideValue = Object.prototype.hasOwnProperty.call(this.parameters.configs, context.key);

      if (!hasConfigOverrideValue && !context.cfgu.lazy) {
        return current;
      }

      const configOverrideValue = this.parameters.configs?.[context.key] ?? '';
      return {
        ...current,
        result: {
          origin: EvaluatedConfigOrigin.ConfigsOverride,
          source: `parameters.configs.${context.key}=${configOverrideValue}`,
          value: configOverrideValue,
        },
      };
    });
  }
```

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 10 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/EvalCommand.ts:174]: done

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 11 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/EvalCommand.ts:40]: done

File: ts/packages/ts/src/commands/commands.test.ts (hunk 8 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/commands.test.ts:333]: done

**Review by RichardAkman — APPROVED** [review]
LGTM 🎖️

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 11 above)
Comment (RonConfigu) [reply ts/packages/ts/src/commands/EvalCommand.ts]: done

File: ts/packages/ts/src/commands/EvalCommand.ts (hunk 9 above)
Comment (rannn505) [reply ts/packages/ts/src/commands/EvalCommand.ts]: @RonConfigu image is broken here

**Review by rannn505 — APPROVED** [review]
@RonConfigu great work
